### PR TITLE
Remove discouraged and deprecated usage of sprintf across the codebase

### PR DESCRIPTION
**Pull request by PatTheMav**
### Description
Replaces all usage of `sprintf` with the safer variant `snprintf` across the codebase.

### Motivation and Context
Usage of `sprintf` has been discouraged for many years now, because the size of the target buffer is not taken into account. `snprintf` should be used instead, which this PR switches to.

`sprintf` is also officially deprecated in macOS 13/Xcode 14.

### How Has This Been Tested?
Compiled and ran OBS on Windows and macOS.

### Types of changes
- Code cleanup (non-breaking change which makes code smaller or more readable)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by tytan652** [comment]
Is the usage of `sizeof(char)` a good idea rather than the size of the char array ?

File: deps/libcaption/src/caption.c [hunk 1]
```diff
@@ -408,10 +408,11 @@ size_t caption_frame_dump_buffer(caption_frame_t* frame, utf8_char_t* buf)
 {
```
Comment (notr1ch) [inline deps/libcaption/src/caption.c:408]: `snprintf` can't be safely used in this function without the size of the buffer being passed as a function.

File: deps/obs-scripting/obs-scripting-python-import.c [hunk 2]
```diff
@@ -65,9 +65,9 @@ bool import_python(const char *python_path)
 
 	char temp[PATH_MAX];
 
-	sprintf(cur_version, VERSION_PATTERN, PY_MAJOR_VERSION_MAX,
-		PY_MINOR_VERSION_MAX);
-	sprintf(temp, FILE_PATTERN, cur_version);
+	snprintf(cur_version, 5, VERSION_PATTERN, PY_MAJOR_VERSION_MAX,
```
Comment (notr1ch) [inline deps/obs-scripting/obs-scripting-python-import.c:68]: Should be `sizeof(cur_version)`, using literals risks errors when one is changed but not the other.

**Review by notr1ch — CHANGES_REQUESTED** [review]
There's a lot of issues with this, almost all of the changes look incorrect. Using the size of individual characters instead of the whole buffer, using literals instead of sizeof and using `snprintf` without knowing the buffer size seem common throughout. My inline comments have only highlighted a few examples but the whole PR seems to have these issues.

**Comment by tytan652** [comment]
You also need to use the return value of `snprintf()` on Linux because otherwise it will/could emit `format-truncation` warnings.

https://github.com/obsproject/obs-studio/pull/5653#discussion_r767311212

File: deps/libcaption/src/caption.c (hunk 1 above)
Comment (PatTheMav) [reply deps/libcaption/src/caption.c:408]: The buffer is initialised with `CAPTION_FRAME_DUMP_BUF_SIZE`, so it should suffice to carry on with that value?

File: UI/window-basic-main-outputs.cpp [hunk 3]
```diff
@@ -1305,7 +1305,7 @@ AdvancedOutput::AdvancedOutput(OBSBasic *main_) : BasicOutputHandler(main_)
 
 	for (int i = 0; i < MAX_AUDIO_MIXES; i++) {
 		char name[9];
-		sprintf(name, "adv_aac%d", i);
+		snprintf(name, sizeof(char), "adv_aac%d", i);
```
[the remark this answers is not in the document]
Comment (PatTheMav) [reply UI/window-basic-main-outputs.cpp:1308]: Yep stupid mistake throughout.

**Comment by PatTheMav** [comment]
> You also need to use the return value of `snprintf()` on Linux because otherwise it could emit `format-truncation` warnings with GCC.
> 
> * [Fix some compillation warnings that appear on Linux #5653 (comment)](https://github.com/obsproject/obs-studio/pull/5653#discussion_r767311212)
> * [Fix some compillation warnings that appear on Linux #5653 (comment)](https://github.com/obsproject/obs-studio/pull/5653#discussion_r848955534)
> * [096ec55](https://github.com/obsproject/obs-studio/commit/096ec55c8370052dd9d0547ad1a670b1bdd08f3f)
> * [ce62240](https://github.com/obsproject/obs-studio/commit/ce62240530b0fc102b799be8f0fa824f7a746300)

`snprintf` is already used without checking for the return value throughout the codebase, don't those throw warnings as well?

**Comment by tytan652** [comment]
> `snprintf` is already used without checking for the return value throughout the codebase, don't those throw warnings as well?

Check Ubuntu CIs, some of them throw a format-truncation warning

**Comment by PatTheMav** [comment]
Fixed the most obvious mistakes - when the char array is created on the stack, `sizeof` is used, otherwise the functions have been amended to pass the size of the char array together with the pointer.

If the char array is initialised with a static preprocessor value, that one is used on both ends.

I'll look at the warnings on Linux next.

File: deps/libcaption/src/caption.c (hunk 1 above)
Comment (notr1ch) [reply deps/libcaption/src/caption.c:408]: That should be OK, I would probably also mark the function as `static` to ensure only callers in caption.c (that presumably are aware of the buffer size) can use it.

**Comment by tytan652** [comment]
No more `-Wformat-truncation`, but three `-Wsign-compare` appeared (3 others are FTL). Some casts and they should be gone.

File: deps/libcaption/src/caption.c (hunk 1 above)
Comment (PatTheMav) [reply deps/libcaption/src/caption.c]: Removed both functions as discussed.

**Comment by PatTheMav** [comment]
Added new commits fixing existing implementations that used literals as buffer length.

File: deps/libcaption/src/caption.c (hunk 1 above)
Comment (RytoEX) [reply deps/libcaption/src/caption.c]: Has this been resolved? Would like to move this forward.

Comment (PatTheMav) [reply deps/libcaption/src/caption.c]: Yeah should be resolved as the corresponding functions have been removed entirely.

File: libobs/util/platform.c [hunk 4]
```diff
@@ -559,7 +559,7 @@ static inline void from_locale(char *buffer)
 double os_strtod(const char *str)
 {
 	char buf[64];
-	snprintf(buf, 64, "%s", str);
+	snprintf(buf, sizeof(buf), "%s", str);
```
Comment (notr1ch) [inline libobs/util/platform.c:562]: Seems like weird use of `snprintf`, `strncpy` / `strncpy_s` (if available) seem more appropriate.

File: plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c [hunk 5]
```diff
@@ -631,10 +631,15 @@ static obs_properties_t *vaapi_properties(void *unused)
 	if (obs_property_list_item_count(list) == 0) {
 		char path[32] = "/dev/dri/renderD1";
 		for (int i = 28;; i++) {
-			sprintf(path, "/dev/dri/renderD1%d", i);
+			snprintf(path, sizeof(path), "/dev/dri/renderD1%d", i);
 			if (access(path, F_OK) == 0) {
 				char card[128] = "Card: ";
```
Comment (notr1ch) [inline plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c:636]: Initializing this seems weird when we immediately overwrite it.

Comment (PatTheMav) [reply plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c]: Changed both `snprintf` invocations to use a `const char` template string and also removed the initial string assignment (as it's immediately overwritten anyway as you mentioned).

File: libobs/util/platform.c (hunk 4 above)
Comment (notr1ch) [reply libobs/util/platform.c]: Unfortunately `strncpy` is unsafe as-is as it doesn't null terminate on overflow. The preferred pattern is e.g. 
```
strncpy (buf, sizeof(buf)-1, str);
buf[sizeof(buf)-1] = 0;
```

File: plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c (hunk 5 above)
Comment (notr1ch) [reply plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c]: I'm personally not a fan of moving the format strings away from the `snprintf` invocation as it adds additional programmer overhead to parse what it's doing, but this solves my main concern.

Comment (PatTheMav) [reply plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c]: I wouldn't be opposed to it, I tried to surmise why it was written that way to begin with and thought the idea was to have a single "template" string that needs to be adjusted, but that train of thought also doesn't make sense.

File: libobs/util/platform.c (hunk 4 above)
Comment (PatTheMav) [reply libobs/util/platform.c]: Right, will change.

**Review by RytoEX — APPROVED** [review]
Got @notr1ch 's approval off-thread.

Will make a mental note that this modifies libcaption and librtmp, which are technically third-party dependencies even though we have them in-tree. Merging shortly.

**Pull request merged**